lib/marin/src/marin/rl/hooks.py

```diff
@@ -0,0 +1,255 @@
+# limitations under the License.
+
+"""Hook system for RolloutWorker to enable flexible callbacks and monitoring.
+
+This module provides a flexible hook system that allows arbitrary callbacks to be
+registered with the RolloutWorker, replacing the previous hardcoded evaluation logic.
+"""
+
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Any
+
+import jax.random as jrandom
+from jax import numpy as jnp
+
+if TYPE_CHECKING:
+    from marin.rl.rollout_worker import RolloutWorker
+
+logger = logging.getLogger("ray")
+
+
+@dataclass
+class HookContext:
+    """Context passed to hooks containing relevant state and utilities."""
+
+    worker: "RolloutWorker"
+    step: int
+    rng: jnp.ndarray
+    lesson_id: str | None = None
+    metadata: dict[str, Any] = field(default_factory=dict)
+
+    def split_rng(self) -> tuple[jnp.ndarray, jnp.ndarray]:
+        """Split the RNG key for use in the hook."""
+        keys = jrandom.split(self.rng)
+        return keys[0], keys[1]
+
+
+class Hook(ABC):
+    """Base class for all hooks that can be registered with RolloutWorker."""
+
+    @abstractmethod
+    def should_run(self, context: HookContext) -> bool:
+        """Determine if this hook should run at the current step.
+
+        Args:
+            context: The current hook context
+
+        Returns:
+            True if the hook should run, False otherwise
+        """
+        pass
+
+    @abstractmethod
+    def run(self, context: HookContext) -> dict[str, Any] | None:
+        """Execute the hook logic.
+
+        Args:
+            context: The current hook context
+
+        Returns:
+            Optional dictionary of metrics or results
+        """
+        pass
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}()"
+
+
+class PeriodicHook(Hook):
+    """Base class for hooks that run periodically based on step count."""
+
+    def __init__(self, frequency: int, start_step: int = 0):
+        """Initialize a periodic hook.
+
+        Args:
+            frequency: Run hook every N steps
+            start_step: First step to start running the hook (default: 0)
+        """
+        self.frequency = frequency
+        self.start_step = start_step
+
+    def should_run(self, context: HookContext) -> bool:
+        """Check if hook should run based on frequency."""
+        return context.step > self.start_step and context.step % self.frequency == 0
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}(frequency={self.frequency}, start_step={self.start_step})"
+
+
+class EvaluationHook(PeriodicHook):
+    """Hook that performs curriculum evaluation - preserves existing behavior."""
+
+    def __init__(
+        self,
+        frequency: int,
+        n_examples: int,
+        eval_type: str = "eval",
+        start_step: int = 0,
+        evaluate_all_lessons: bool = False,
```

Comment:
we tend to prefer to have separate calls/classes instead of boolean dispatch.  e.g.

evaluate_curriculum and evaluate_sample

vs evaluate(curriculum_or_sample)

So here we might have an EvaluateCurriculum and EvaluateLesson hooks, that also cleans up the configuration. Once you do that, it's probably less necessary to docstring as aggressively as well, e.g. you could just have:

class EvaluateCurriculumHook(PeriodicHook):
   """Evaluate all lessons in the curriculum periodically"""
  def run(self, context):
     return worker.evaluate..